proxy/server/server.go

```diff
@@ -274,6 +274,10 @@ func dispatch(ctx context.Context, requestChan chan *pb.ProxyRequest, replyChan
 				streamSet.ClientCloseAll()
 				return nil
 			}
+			// Peer information might not be properly populated until rpcauth
+			// evaluates the initial received message, so let's grab fresh
+			// peer information when we know we've gotten at least one message.
+			ctx = rpcauth.AddPeerToContext(ctx, rpcauth.PeerInputFromContext(stream.Context()))
```

Comment:
I have a vague worry about this generating many, many contexts. Walking the codepath here, AddPeerToContext calls WithContext, which creates a new context object on each call. It seems 'cheap' to add a boolean here to track whether we've done this dance once, and avoid doing it every time we receive a message on the channel

Reply:
Good call, added a boolean. We should have good guarantees that the authz hook has run before any message is received by this channel.